File: server/github.js
import fetch from "node-fetch";
import fs from "fs";

export default class GitHub {
    URL = "https://api.github.com/repos/REDasmOrg/REDasm";
    ARTIFACTS = "artifacts.json";
    RELEASES = "releases.json";

    async getReleases() { return await this._check("releases", this.RELEASES); }
    async getArtifacts() { return await this._check("actions/artifacts", this.ARTIFACTS); }

    async _req(q, filename) {
        let response = await fetch(`${this.URL}/${q}`);

        if(response.ok) {
            response = await response.json();
            fs.writeFileSync(filename, JSON.stringify(response));
        }

        return this._read(filename);
    }

    async _check(q, filename) {
        let res = null;
        if(this._isExpired(filename)) res = await this._req(q, filename);
        else res = this._read(filename);
        return res;
    }

    _read(filename) {
        try { return JSON.parse(fs.readFileSync(filename)); }
        catch { return null; }
    }

    _isExpired(filename) {
        if(!fs.existsSync(filename)) return true;
        const { birthtime } = fs.statSync(filename);
        let days = Math.round(Math.abs(new Date() - birthtime) / 36e5);
        return days > 0; // >= 1 day = expired
    }
}

File: server/src/github.nim
import httpclient, strformat, json, times, os

const GITHUB_URL = "https://api.github.com/repos/REDasmOrg/REDasm"
const ARTIFACTS_FILE = "artifacts.json"
const RELEASES_FILE = "releases.json"

proc isExpired(filename: string): bool =
  if not fileExists(filename): return true

  let now = getTime()
  let fi = getFileInfo(filename);
  return (now - fi.lastWriteTime).inDays > 0;

proc read(filename: string): JsonNode =
  try:
    return parseJson(readFile(filename))
  except:
    return newJNull()

proc request(q: string, filename: string): JsonNode =
  let client = newHttpClient()

  try:
    let s = client.getContent(fmt"{GITHUB_URL}/{q}")
    writeFile(filename, s)
    return parseJson(s)
  except:
    return newJNull()

proc check(q: string, filename: string): JsonNode =
  if isExpired(filename):
    return request(q, filename)

  return read(filename)

proc getReleases*(): JsonNode =
  return check("releases", RELEASES_FILE);

proc getArtifacts*(): JsonNode =
  return check("actions/artifacts", ARTIFACTS_FILE);

